操作符 API 翻译：dematerialize -> do + materialize

SangKa opened the pull request:
完成下列操作符的 API 翻译：
  - [x] dematerialize 
  - [x] distinct
  - [x] distinctUntilChanged 
  - [x] distinctUntilKeyChanged 
  - [x] do
  - [x] materialize

SangKa committed:
docs(materialize): translate api materialize.ts<100%>

SangKa committed:
docs(dematerialize): translate api dematerialize.ts<100%>

SangKa committed:
docs(distinct): translate api distinct.ts<100%>

SangKa committed:
docs(distinctUntilChanged): translate api distinctUntilChanged.ts<100%>

SangKa committed:
docs(distinctUntilKeyChanged): translate api distinctUntilKeyChanged.ts<100%>

SangKa committed:
docs(do): translate api do.ts<100%>

SangKa commented:
@moreJs 请 review

moreJs commented:
@SangKa 好的，我的pr我明天提交给你。

moreJs commented on src/operator/dematerialize.ts:
```diff
@@ -4,23 +4,21 @@ import { Subscriber } from '../Subscriber';
- * `error`. Such Observable is the output of a `materialize` operation. Those
- * notifications are then unwrapped using the metadata they contain, and emitted
- * as `next`, `error`, and `complete` on the output Observable.
+ * `dematerialize` 被假定为操作的 Observable 只会将 {@link Notification} 对象
```
这句读起来不太通顺，我理解应该是这样：`dematerialize`被假定用来操作只发送`next`不发送`error`的Observable

SangKa replied:
同意

moreJs commented on src/operator/dematerialize.ts:
```diff
@@ -4,23 +4,21 @@ import { Subscriber } from '../Subscriber';
+ * 与 {@link materialize} 结合来使用此操作符。
  *
- * @example <caption>Convert an Observable of Notifications to an actual Observable</caption>
+ * @example <caption>将 Notifications 的 Observable 转换成实际的 Observable</caption>
```
'Notifications的Observable' 不太通，我理解：将Notification型的Observable转化为真正的Observable

SangKa replied:
前半句同意，后半句 `actual` 翻译为真正的并不认同，因为本身就是 Observable ，真正的不能很好的达意。

moreJs replied:
对，他们属于继承链的不同层次，这块我不知道怎么表达更好点。

moreJs commented on src/operator/dematerialize.ts:
```diff
@@ -30,16 +28,16 @@ import { Notification } from '../Notification';
- * @return {Observable} An Observable that emits items and notifications
- * embedded in Notification objects emitted by the source Observable.
+ * @return {Observable} Observable 会发出数据项和通知，它们是由源 Observable 所发出
+ * 并且包装在 Notification 对象之中的。
```
'embedded in' 直译为'嵌入'会不会更好？

SangKa replied:
可以 但感觉不会更好。

moreJs commented on src/operator/distinct.ts:
```diff
@@ -8,25 +8,23 @@ import { subscribeToResult } from '../util/subscribeToResult';
- * Returns an Observable that emits all items emitted by the source Observable that are distinct by comparison from previous items.
+ * 返回 Observable，它发出源 Observable 发出的所有与之前所有项都不相同的项。
+ * 
+ * 如果提供了 keySelector 函数，那么它会将源 Observable 的每个值都投射成一个新的值，这个值会用来检查是否与前一个投射值相等。如果没有提供 
```
‘与前一个投射值相等’不太准确，我理解是:`与先前投射的值是否相等`

SangKa replied:
同意 这里确实翻译不准确

moreJs commented on src/operator/distinct.ts:
```diff
@@ -8,25 +8,23 @@ import { subscribeToResult } from '../util/subscribeToResult';
+ * 在支持 `Set` 的 JavaScript 运行时中，此操作符会使用 `Set` 来提升不同值检查的性能。
  *
- * In JavaScript runtimes that support `Set`, this operator will use a `Set` to improve performance of the distinct value checking.
+ * 在其他运行时中，此操作符会使用 `Set` 的最小化实现，此实现在底层依赖于 `Array` 和 `indexOf`，因为要检查更多的值来进行区分，所以性能会降低。
```
‘最小化实现’ 是否没有 '简化版' 更好点？

SangKa replied:
简化版无法充分表达 minimal 的含义

SangKa committed:
docs(operators): polish translation of operators

operator list:

- dematerialize
- distinct
- distinctUntilChanged
- distinctUntilKeyChanged
- materialize

moreJs commented on src/operator/distinct.ts:
```diff
@@ -8,25 +8,23 @@ import { subscribeToResult } from '../util/subscribeToResult';
 
 /**
- * Returns an Observable that emits all items emitted by the source Observable that are distinct by comparison from previous items.
+ * 返回 Observable，它发出由源 Observable 所发出的所有与之前所有项都不相同的项。
```
两个 “所有” 有点重复 啰嗦了吧？ 建议去掉一个

SangKa replied:
同意

moreJs commented on src/operator/distinct.ts:
```diff
@@ -39,16 +37,16 @@ import { ISet, Set } from '../util/Set';
- * @param {function} [keySelector] Optional function to select which value you want to check as distinct.
- * @param {Observable} [flushes] Optional Observable for flushing the internal HashSet of the operator.
- * @return {Observable} An Observable that emits items from the source Observable with distinct values.
+ * @param {function} [keySelector] 可选函数，用来选择要检查哪个值是不同的。
```
这句话，读起来，感觉不太通顺。

SangKa replied:
那给个建议呢

SangKa replied:
可选函数，用来选择某个键的值以检查是否是不同的。

moreJs commented:
@SangKa  review 完成。

SangKa committed:
docs(distinct): polish translation
